code/glTF2AssetWriter.inl

```diff
@@ -196,51 +206,158 @@ namespace glTF2 {
+            Value arr;
+            obj.AddMember(StringRef(propName), MakeValue(arr, prop, al), al);
+        }
+
+        template<size_t N>
+        inline void WriteVec(Value& obj, float(&prop)[N], const char* propName, float(&defaultVal)[N], MemoryPoolAllocator<>& al)
+        {
+            if (!std::equal(std::begin(prop), std::end(prop), std::begin(defaultVal))) {
+                WriteVec(obj, prop, propName, al);
             }
         }
+
+        inline void WriteFloat(Value& obj, float prop, const char* propName, MemoryPoolAllocator<>& al)
+        {
+            Value num;
+            obj.AddMember(StringRef(propName), MakeValue(num, prop, al), al);
+        }
     }
 
     inline void Write(Value& obj, Material& m, AssetWriter& w)
     {
-        if (m.transparent) {
-            obj.AddMember("alphaMode", "BLEND", w.mAl);
+        Value pbrMetallicRoughness;
+        pbrMetallicRoughness.SetObject();
+        {
+            WriteTex(pbrMetallicRoughness, m.pbrMetallicRoughness.baseColorTexture, "baseColorTexture", w.mAl);
+            WriteTex(pbrMetallicRoughness, m.pbrMetallicRoughness.metallicRoughnessTexture, "metallicRoughnessTexture", w.mAl);
+
+            //@TODO: define this as a constant?
+            vec4 defaultEmissiveFactor = {1, 1, 1, 1};
+            WriteVec(pbrMetallicRoughness, m.pbrMetallicRoughness.baseColorFactor, "baseColorFactor", defaultEmissiveFactor, w.mAl);
+
+            if (m.pbrMetallicRoughness.metallicFactor != 1) {
+                WriteFloat(pbrMetallicRoughness, m.pbrMetallicRoughness.metallicFactor, "metallicFactor", w.mAl);
+            }
+
+            if (m.pbrMetallicRoughness.roughnessFactor != 1) {
+                WriteFloat(pbrMetallicRoughness, m.pbrMetallicRoughness.roughnessFactor, "roughnessFactor", w.mAl);
+            }
         }
 
-        Value v;
-        v.SetObject();
-        {
-            if (m.transparent && !m.diffuse.texture) {
-                m.diffuse.color[3] = m.transparency;
+        if (!pbrMetallicRoughness.ObjectEmpty()) {
+            obj.AddMember("pbrMetallicRoughness", pbrMetallicRoughness, w.mAl);
+        }
+
+        WriteTex(obj, m.normalTexture, "normalTexture", w.mAl);
+        WriteTex(obj, m.emissiveTexture, "emissiveTexture", w.mAl);
+        WriteTex(obj, m.occlusionTexture, "occlusionTexture", w.mAl);
+
+        //@TODO: define this as a constant?
+        vec3 defaultEmissiveFactor = {0, 0, 0};
+        WriteVec(obj, m.emissiveFactor, "emissiveFactor", defaultEmissiveFactor, w.mAl);
+
+        if (m.alphaCutoff != 0.5) {
+            WriteFloat(obj, m.alphaCutoff, "alphaCutoff", w.mAl);
+        }
+
+        if (m.alphaMode != "OPAQUE") {
+            obj.AddMember("alphaMode", m.alphaMode, w.mAl);
+        }
+
+        if (m.doubleSided) {
+            obj.AddMember("doubleSided", m.doubleSided, w.mAl);
+        }
+
+        Value exts;
+        exts.SetObject();
+
+        if (m.pbrSpecularGlossiness.on) {
```

Comment:
Is `.on` a valid form of flagging whether pbrSpecularGlossiness is enabled? Or should I use/instantiate a pointer, and check whether or not it is null?

Reply:
Unfortunately C++ doesn't have an idiomatic way to do this--what it really needs is an option type. My personal preference, which doesn't seem to be the norm, is to use a `std::unique_ptr<T>`, which is an object that wraps a `T*`--pointer to `T`--and automatically `delete`s it, so you don't have to worry so much about memory leaks. Another option is to have a `pbrSpecularGlossinessOn` flag on the material itself. Your solution seems at least as clean as those two, so it's up to you.

Reply:
Your preference seems like a good idea, but is a completely foreign concept to me. If you think it's better, feel free to make the changes.

Reply:
Nah, I think what you've done is much simpler and cleaner. Using pointers as an option type is a bit of a non-obvious abuse of notation, even though they have the same semantics.

Reply:
Ooh, just noticed that there's a simple `Nullable` struct already defined in `glTF2Asset.h`: https://github.com/Matter-and-Form/assimp/blob/26c3bbfc0ffcae368c5eb9229ee67c35875a6065/code/glTF2Asset.h#L372.
So maybe you could make the `pbrSpecularGlossiness` a `Nullable<PbrSpecularGlossiness>` and omit the `on` property?

Reply:
How do I enable/disable (null / not-null) it in the code?

Reply:
There's an `isPresent` flag on the `Nullable`. If you use the constructor that takes no arguments, it's automatically set to `false`; if you pass in a value to the constructor, it's `true`.